vsock: support header and data on a single descriptor

stefano-garzarella opened the pull request:
### Summary of the PR

Starting from Linux 6.2 the virtio-vsock driver can use a single descriptor for both header and data:
https://lore.kernel.org/lkml/20221215043645.3545127-1-bobby.eshleman@bytedance.com/

So with this modification the virtio-vsock device can support header and data on a single descriptor or on two. This is just a Linux implementation detail though, for the spec it could be multiple descriptors as well.

For now let's add only the single descriptor support, since it doesn't involve an API change, but in the future we should change the API and support an array of VolatileSlices as suggested by Laura here: https://github.com/rust-vmm/vm-virtio/issues/204#issuecomment-1332511007

Let's make the tests work with this change (mainly by always using PKT_HEADER_SIZE for the header descriptor) and adding new tests to cover the single descriptor scenario.

Fixes: #204


### Requirements

Before submitting your PR, please make sure you addressed the following
requirements:

- [x] All commits in this PR are signed (with `git commit -s`), and the commit
  message has max 60 characters for the summary and max 75 characters for each
  description line.
- [x] All added/changed functionality has a corresponding unit/integration
  test.
- [x] All added/changed public-facing functionality has entries in the "Upcoming 
  Release" section of CHANGELOG.md (if no such section exists, please create one).
- [x] Any newly added `unsafe` code is properly documented.

lauralt commented:
> About the CI failure on the commit message, should I remove the links? 

No, we can merge the PR with admin rights when it'll be the time.

andreeaflorescu commented:
@stefano-garzarella The tests that we are now using for generating the fuzz input are failing, can you take a look?
https://buildkite.com/rust-vmm/vm-virtio-ci/builds/728#01852a4d-1e33-4d66-885b-006d851b7344

stefano-garzarella commented:
> @stefano-garzarella The tests that we are now using for generating the fuzz input are failing, can you take a look? https://buildkite.com/rust-vmm/vm-virtio-ci/builds/728#01852a4d-1e33-4d66-885b-006d851b7344

Ooops, I forgot to adjust the descriptors in the fuzz subfolder.

andreeaflorescu commented:
@stefano-garzarella if you do a rebase on top of main the `cargo audit` test should now pass, and we shouldn't see any more failures. Besides the commit message test all the other tests should pass to merge this PR.

andreeaflorescu commented on crates/devices/virtio-vsock/src/packet.rs:
```diff
@@ -460,19 +460,30 @@ impl<'a, B: BitmapSlice> VsockPacket<'a, B> {
+            mem.get_slice(
+                chain_head
+                    .addr()
+                    .checked_add(PKT_HEADER_SIZE.try_into().unwrap())
```
nit: Since we are doing a `try_into()` followed by an `unwrap` I think we can just use `PKT_HEADER_SIZE as u64`. This is safe since the PKT_HEADER_SIZE is a usize. Unwraps tend to make you think that something potentially dangerous us going on 😆

stefano-garzarella replied:
ack, I'll change it!

andreeaflorescu commented on crates/devices/virtio-vsock/src/packet.rs:
```diff
@@ -460,19 +460,30 @@ impl<'a, B: BitmapSlice> VsockPacket<'a, B> {
-        let data_desc = desc_chain.next().ok_or(Error::DescriptorChainTooShort)?;
+        // Starting from Linux 6.2 the virtio-vsock driver can use a single descriptor for both
+        // header and data.
+        let data_slice = if chain_head.len() >= PKT_HEADER_SIZE as u32 + pkt.len() {
```
The addition operation can overflow. Unfortunately not having a maximum (and reasonable) max packet size means that a wrongly initialized value can lead to an overflow here in case you're also dealing with a malicious user.

andreeaflorescu replied:
To reproduce:
```rust
        // On the TX path, it is allowed to not have a data descriptor.
        let v = vec![Descriptor::new(0x10_0000, PKT_HEADER_SIZE as u32, 0, 0)];
        let mut chain = queue.build_desc_chain(&v[..1]).unwrap();

        let header = PacketHeader {
            src_cid: SRC_CID.into(),
            dst_cid: DST_CID.into(),
            src_port: SRC_PORT.into(),
            dst_port: DST_PORT.into(),
            len: u32::MAX.into(),
            type_: 0.into(),
            op: 0.into(),
            flags: 0.into(),
            buf_alloc: 0.into(),
            fwd_cnt: 0.into(),
        };
        mem.write_obj(header, GuestAddress(0x10_0000)).unwrap();

        let packet = VsockPacket::from_tx_virtq_chain(&mem, &mut chain, u32::MAX).unwrap();
```

lauralt replied:
> The addition operation can overflow. Unfortunately not having a maximum (and reasonable) max packet size means that a wrongly initialized value can lead to an overflow here in case you're also dealing with a malicious user.

Which reminds me that I should at some point resume work (addressing comments actually) on https://lists.oasis-open.org/archives/virtio-comment/202206/msg00030.html 😄 .

stefano-garzarella replied:
Okay, I'll change the check to `if chain_head.len() - PKT_HEADER_SIZE as u32 >= pkt.len()`, so every potential overflow will go to the `else` branch and handled as before this patch.

Is that okay for you?

andreeaflorescu replied:
I think this will still make the fuzzer fail because overflows are panics in debug mode, and if I am not mistaken you can still have an underflow? Will double check.

stefano-garzarella replied:
At that point, we already checked that `chain_head.len()` is  `>= PKT_HEADER_SIZE`, so I think we can't have underflow or overflow. Or am I forgetting something? (My mind is already in the holiday mood 🎄).

andreeaflorescu replied:
Yeah, I know what you mean, same here 🤶. Let me take a better look after some more ☕

andreeaflorescu replied:
This overflow is indeed no longer possible. Resolving this conversation.

andreeaflorescu commented on crates/devices/virtio-vsock/src/packet.rs:
```diff
@@ -583,27 +594,39 @@ impl<'a, B: BitmapSlice> VsockPacket<'a, B> {
-        }
+        // Starting from Linux 6.2 the virtio-vsock driver can use a single descriptor for both
+        // header and data.
+        let data_slice = if chain_head.len() as usize > PKT_HEADER_SIZE {
```
We typically want to compare stuff.len() to the PKT_HEADER_SIZE. The `len` function uses an atypical return type for a `len` function which makes these conversions required. Maybe we can open an issue to fix this separately if this is not on purpose.

CC: @lauralt

lauralt replied:
I think what we wanted here was to be consistent with the spec (len is defined as le32 there). We can also update the method to return an usize, but I would be in favor of keeping it as defined in the spec.

stefano-garzarella replied:
I don't have a strong opinion on that, but I agree that we should change it in another PR.

andreeaflorescu replied:
Ok, it looks like it's on purpose like this, so I will resolve this.

andreeaflorescu reviewed (changes requested):
Didn't look over all the code, but I think it needs some adjustments. Will take a look at the rest tomorrow.

andreeaflorescu commented:
@stefano-garzarella I created a longer running fuzzing pipeline for PRs that runs fuzzing for 1 day, but doesn't block the merge of PRs. Do you mind if I do an update branch on your PR to trigger the run?

stefano-garzarella commented:
> @stefano-garzarella I created a longer running fuzzing pipeline for PRs that runs fuzzing for 1 day, but doesn't block the merge of PRs. Do you mind if I do an update branch on your PR to trigger the run?

It's fine, but the current branch doesn't include the change that I proposed to avoid the overflow.
Let me push it and IIUC the 1day pipeline will start, right?

stefano-garzarella commented:
> > @stefano-garzarella I created a longer running fuzzing pipeline for PRs that runs fuzzing for 1 day, but doesn't block the merge of PRs. Do you mind if I do an update branch on your PR to trigger the run?
> 
> It's fine, but the current branch doesn't include the change that I proposed to avoid the overflow. Let me push it and IIUC the 1day pipeline will start, right?

ah no, I thought was on main. So I'll send the new version and then you can update it.

andreeaflorescu commented:
> > @stefano-garzarella I created a longer running fuzzing pipeline for PRs that runs fuzzing for 1 day, but doesn't block the merge of PRs. Do you mind if I do an update branch on your PR to trigger the run?
> 
> It's fine, but the current branch doesn't include the change that I proposed to avoid the overflow. Let me push it and IIUC the 1day pipeline will start, right?

Yap, that's right! Once it is pushed once, then we can also retrigger it from github. But since the webhook was yet never called, GitHub doesn't know of its existence.

andreeaflorescu commented on crates/devices/virtio-vsock/src/packet.rs:
```diff
@@ -867,6 +920,33 @@ mod tests {
+            VRING_DESC_F_WRITE as u16,
+            0,
+        )];
+        let mut chain = queue.build_desc_chain(&v[..1]).unwrap();
```
nit: you can pass `&v` directly.

stefano-garzarella replied:
oh right, copy&past from other tests, I think we can pass `&v` everywhere.
I can send a cleanup on top of this.

andreeaflorescu replied:
Yes, right. I noticed that in the other tests as well, but didn't want to burden you to change it all 😆. On the other hand I also don't want to propagate the mistake to other places.

stefano-garzarella replied:
No problem at all.
I'll fix the tests added here and fix the other in a separate commit.
Should I integrate that commit in this PR, or better a new one?

andreeaflorescu replied:
Works either way as long as it's a separate commit. Whichever makes it easier for you.

stefano-garzarella replied:
Okay, I'll include here, less noise for others ;-)

andreeaflorescu reviewed (approved):
We can probably get rid of some of the duplication in the tx and rx path, but I don't think that makes a lot of sense since we want to do a more generic refactoring anyway.

LGTM.

lauralt commented:
[This documentation](https://github.com/rust-vmm/vm-virtio/blob/main/crates/devices/virtio-vsock/src/packet.rs#L15) needs to be updated.

lauralt commented:
We should probably also add a CHANGELOG entry for this, even if it's not an API change.

stefano-garzarella commented:
new version:
- fixed [documentation](https://github.com/rust-vmm/vm-virtio/blob/main/crates/devices/virtio-vsock/src/packet.rs#L15)
- added CHANGELOG entries

lauralt commented on crates/devices/virtio-vsock/src/packet.rs:
```diff
@@ -583,27 +595,39 @@ impl<'a, B: BitmapSlice> VsockPacket<'a, B> {
-        }
+        // Starting from Linux 6.2 the virtio-vsock driver can use a single descriptor for both
+        // header and data.
+        let data_slice = if chain_head.len() as usize > PKT_HEADER_SIZE {
```
I'm wondering if, for the time being (until we change to multiple `VolatileSlice`s), we should also check here if there's no next descriptor in the chain. If there is another descriptor in the chain, we keep the same behaviour as before, ignore the bytes after the header size in the first descriptor, and use only the second one. This kind of translates to: what faulty behaviour do we want to allow?

stefano-garzarella replied:
It makes sense to me in the sense of keeping the same behavior as before, but in my opinion we should not have ignored the bytes after the header in the first descriptor. In theory those are part of the payload.

So for now I agree to try to change as little as possible and check both the length and the not having a next. (So I can restore all the tests I changed as well).

In the future let's fix that and handle the various descriptors simply as a sequence of bytes.

What do you think?

lauralt replied:
> It makes sense to me in the sense of keeping the same behavior as before, but in my opinion we should not have ignored the bytes after the header in the first descriptor. In theory those are part of the payload.
> 
> So for now I agree to try to change as little as possible and check both the length and the not having a next. (So I can restore all the tests I changed as well).
> 
> In the future let's fix that and handle the various descriptors simply as a sequence of bytes.
> 
> What do you think?

Yup, I agree. I don't think we can achieve that without the multiple VolatileSlices API change, so we should prioritise this refactoring.

lauralt replied:
I will open a separate issue for that, I see your PR will close the one opened by you, which is okay I think.

lauralt reviewed (approved):
Do we publish this as a patch release, in that case we should follow [this guide](https://github.com/rust-vmm/community/blob/main/docs/maintainers/crate_release.md#patch-release) which involves creating a new branch etc.

andreeaflorescu commented:
> Do we publish this as a patch release, in that case we should follow [this guide](https://github.com/rust-vmm/community/blob/main/docs/maintainers/crate_release.md#patch-release) which involves creating a new branch etc.

I don't think we need a patch release on a new branch unless we want to backport this fix to previous releases. We can just do patch releases from main as well as long as the history is linear, which is what I think is the case here.

stefano-garzarella commented:
new version:
- checked also that `chain_head.next()` is false as @lauralt suggested

lauralt commented on crates/devices/virtio-vsock/src/packet.rs:
```diff
@@ -460,22 +461,34 @@ impl<'a, B: BitmapSlice> VsockPacket<'a, B> {
+                )
+                .map_err(Error::InvalidMemoryAccess)?
+            } else {
+                let data_desc = desc_chain.next().ok_or(Error::DescriptorChainTooShort)?;
```
nit: I would check before is the next bit of the head is set (with has_next()).

stefano-garzarella replied:
Okay, I think it was pre-existing since I just moved the code in the else branch, but since I'm touching this code, make sense to update it ;-)
